### ISSUE TITLE
Frozen sc training error issue

### ISSUE THREAD

**Issue by miasakachenmo** [davidnvq/grit#13]
Hello! When I try your frozen mode as:
```
python train_caption.py exp.name=caption_4ds model.detector.checkpoint=ckpts/detector_checkpoint_vg.pth \
exp.ngpus_per_node=1 \
exp.world_size=1 \
optimizer.freezing_xe_epochs=10 \
optimizer.freezing_sc_epochs=10 \
optimizer.finetune_xe_epochs=0 \
optimizer.finetune_sc_epochs=0
```
XE stage runs well, error occurs in frozen fc stage
```
caption_engine.py:train_sc
with tqdm(desc='Epoch %d - train' % epoch, unit='it', total=len(dataloaders['train_dict'])) as pbar:
    for it, batch in enumerate(dataloaders['train_dict']):
        if 'samples' in batch:
            b_s = batch['samples'].tensors.shape[0]
        elif 'vis_feat' in batch:
            b_s = batch['vis_feat'].shape[0]
```
the exception says: batch['samples'] :'dict' object has no attribute 'tensors'
It seems that the batch['samples'] should be NestedTensor

But actually in frozen sc mode, the dataloader act like this:
```
coco.py:
if self.img_field.use_hdf5_feat:
    samples = {}
    if self.img_field.use_gri_feat:
        samples['gri_feat'] = torch.stack([im['gri_feat'] for im in imgs]).to(self.device)
        samples['gri_mask'] = torch.stack([im['gri_mask'] for im in imgs]).to(self.device)
    if self.img_field.use_reg_feat:
        samples['reg_feat'] = torch.stack([im['reg_feat'] for im in imgs]).to(self.device)
        samples['reg_mask'] = torch.stack([im['reg_mask'] for im in imgs]).to(self.device)
    outputs['samples'] = samples
else:
    outputs['samples'] = nested_tensor_from_tensor_list(imgs).to(self.device)
```
so it returns tensors in hdf5 mode but frozen sc want a Nested tensor?
Or am I wrong in somewhere when running the code?
Thank you for your attention.

**Comment by miasakachenmo** [comment]
After change train_sc to:
```
if 'samples' in batch:
    #if bathc['samples'] is NestedTensor
    if isinstance(batch['samples'], NestedTensor):
        b_s = batch['samples'].tensors.shape[0]
    else:
        b_s = batch['samples']['gri_feat'].shape[0]
elif 'vis_feat' in batch:
    b_s = batch['vis_feat'].shape[0]
```
It works since the program just need batch_size here

**Comment by miasakachenmo** [comment]
Another question:
Can you provide an training log to let us know if our model is training well during the training process?

**Comment by davidnvq** [comment]
Thanks for fixing the reported bug. 

> Can you provide an training log to let us know if our model is training well during the training process?

Using the default config: finetuning the entire model with pretrained detector on 4 datasets, you will obtain the `result.txt` as follows (one of the seeds). Specifically, you only need to set the `seed = 42` to the default config.

```
cat result.txt
Epoch 0: valid scores: {'BLEU': [0.7539302065464176, 0.5964850610373361, 0.4540914149883598, 0.3398709580083359], 'METEOR': 0.26058321438443544, 'ROUGE': 0.5559232010500881, 'CIDEr': 1.0509189515003243}
Epoch 0: test scores: {'BLEU': [0.7539772776964557, 0.5950121456088614, 0.4529896824562538, 0.33881310331024816], 'METEOR': 0.26086421742904187, 'ROUGE': 0.5563483971098572, 'CIDEr': 1.0570154749979377}
Epoch 1: test scores: {'BLEU': [0.7765168310885853, 0.621626713573914, 0.4831692889447988, 0.3726892383086823], 'METEOR': 0.2785012305916916, 'ROUGE': 0.5748451695313205, 'CIDEr': 1.1813618693898293}
Epoch 1: valid scores: {'BLEU': [0.7802547027810786, 0.6266154688777543, 0.48761826736594, 0.37627016318160994], 'METEOR': 0.2799466985163502, 'ROUGE': 0.5776183315614145, 'CIDEr': 1.1777197852223726}
Epoch 2: valid scores: {'BLEU': [0.7860073001050624, 0.6352163821456382, 0.4991021284322571, 0.3896647025962745], 'METEOR': 0.28612635636292105, 'ROUGE': 0.5846688378962767, 'CIDEr': 1.2229133971831447}
Epoch 2: test scores: {'BLEU': [0.7828277078724676, 0.6292161129433145, 0.4919464827254148, 0.3818513474463031], 'METEOR': 0.28487966803466663, 'ROUGE': 0.581130634785704, 'CIDEr': 1.2205020461533056}
Epoch 3: valid scores: {'BLEU': [0.7915261126312594, 0.6413430552239501, 0.5036933210640915, 0.3925653082454366], 'METEOR': 0.2897685482391987, 'ROUGE': 0.5898614712021756, 'CIDEr': 1.238774705397189}
Epoch 3: test scores: {'BLEU': [0.7894787137534327, 0.6342761420314864, 0.4951032649710683, 0.3836893800402813], 'METEOR': 0.28813049526048756, 'ROUGE': 0.5860630653694614, 'CIDEr': 1.2419735127278222}
Epoch 4: test scores: {'BLEU': [0.7920013155569735, 0.6426746947681649, 0.5086630486101409, 0.4000228764515629], 'METEOR': 0.2930986219985243, 'ROUGE': 0.5907662237660707, 'CIDEr': 1.2741988739804082}
Epoch 4: valid scores: {'BLEU': [0.7939936411250952, 0.6442270362570511, 0.5085463295718785, 0.39867903045872266], 'METEOR': 0.2921640240546441, 'ROUGE': 0.5922226290019371, 'CIDEr': 1.260800922777133}
Epoch 5: test scores: {'BLEU': [0.794122298070822, 0.6442399149242897, 0.5095045769566264, 0.4004634296436291], 'METEOR': 0.2947609379633544, 'ROUGE': 0.5942193508972228, 'CIDEr': 1.2832084017450183}
Epoch 5: valid scores: {'BLEU': [0.7985186815004045, 0.6475777224206791, 0.510459452981429, 0.3996505982813784], 'METEOR': 0.2947795292801748, 'ROUGE': 0.5954523392325087, 'CIDEr': 1.265085408335824}
Epoch 6: test scores: {'BLEU': [0.7925353090489129, 0.639374475504484, 0.5050656391874848, 0.39648232551317564], 'METEOR': 0.2955932931154769, 'ROUGE': 0.590542960924995, 'CIDEr': 1.2831039390309849}
Epoch 6: valid scores: {'BLEU': [0.792876924067736, 0.6400784563294283, 0.5061817333554944, 0.39739665056743584], 'METEOR': 0.2955194409183982, 'ROUGE': 0.5914359276946326, 'CIDEr': 1.260464195382476}
Epoch 7: test scores: {'BLEU': [0.7973263870460862, 0.647127518525604, 0.512179643603503, 0.40405251134711284], 'METEOR': 0.29663001116118376, 'ROUGE': 0.5957286526028304, 'CIDEr': 1.3004806036061252}
Epoch 7: valid scores: {'BLEU': [0.7932685654699972, 0.6427967609346499, 0.5073358273797164, 0.39808382205728127], 'METEOR': 0.2943443606873238, 'ROUGE': 0.5917395134410733, 'CIDEr': 1.266374364328324}
Epoch 8: test scores: {'BLEU': [0.8013335264976654, 0.6509646356201589, 0.5148391237124121, 0.40384169481960924], 'METEOR': 0.2975485007961343, 'ROUGE': 0.5960558566070715, 'CIDEr': 1.3026576481041294}
Epoch 8: valid scores: {'BLEU': [0.7996523755006688, 0.6504922976049965, 0.5150792831135323, 0.40471128366868353], 'METEOR': 0.29664606392548537, 'ROUGE': 0.5945415114514051, 'CIDEr': 1.2819505722442046}
Epoch 9: test scores: {'BLEU': [0.7970567017355554, 0.6463949468892651, 0.5112789737535356, 0.4017031612727288], 'METEOR': 0.2973227606936456, 'ROUGE': 0.5942380167964469, 'CIDEr': 1.2990942671806631}
Epoch 9: valid scores: {'BLEU': [0.7966697468002908, 0.6468535353714818, 0.5114752272114751, 0.40223141972119414], 'METEOR': 0.29721832733306075, 'ROUGE': 0.5947410492638745, 'CIDEr': 1.2786217738926917}
Epoch 10: test scores: {'BLEU': [0.8361024192335684, 0.691186075214889, 0.5493720520695162, 0.4278083856490902], 'METEOR': 0.30479439029610017, 'ROUGE': 0.6094661443329562, 'CIDEr': 1.4062434121231884}
Epoch 10: valid scores: {'BLEU': [0.8388904951984599, 0.6925412007037333, 0.5479590305139721, 0.424831181708297], 'METEOR': 0.30354400045524615, 'ROUGE': 0.609063180379453, 'CIDEr': 1.3918860391618548}
Epoch 11: valid scores: {'BLEU': [0.833247402449971, 0.6913725373284553, 0.5480488759761322, 0.42373522293666016], 'METEOR': 0.29994942822665616, 'ROUGE': 0.6098326052855442, 'CIDEr': 1.3794068053587754}
Epoch 11: test scores: {'BLEU': [0.8308709908643401, 0.6881580829623243, 0.548426455984043, 0.4283832453237107], 'METEOR': 0.3005035307160139, 'ROUGE': 0.6091794503965545, 'CIDEr': 1.3955938211728662}
Epoch 12: test scores: {'BLEU': [0.8352954305542933, 0.6926138800228683, 0.5521701943867067, 0.4307247979829361], 'METEOR': 0.3007190772499965, 'ROUGE': 0.6086274140584754, 'CIDEr': 1.4034260975901944}
Epoch 12: valid scores: {'BLEU': [0.8372521316215876, 0.6950327997723531, 0.5519987899414815, 0.42882394451677525], 'METEOR': 0.3011119806832263, 'ROUGE': 0.6089989797379655, 'CIDEr': 1.391657120534352}
Epoch 13: test scores: {'BLEU': [0.8352687700840109, 0.6911561534953273, 0.5491909417245747, 0.42677781205935983], 'METEOR': 0.3008918959768078, 'ROUGE': 0.6089066093769083, 'CIDEr': 1.4011814188287004}
Epoch 13: valid scores: {'BLEU': [0.836489155811366, 0.6937619448377661, 0.5503156864925394, 0.4273261729669541], 'METEOR': 0.30064283562191574, 'ROUGE': 0.6092599638897486, 'CIDEr': 1.3919854962577367}
Epoch 14: test scores: {'BLEU': [0.8396184624281761, 0.6933573832170274, 0.5499467597706865, 0.4274293637633909], 'METEOR': 0.30081124346958704, 'ROUGE': 0.6095786463789528, 'CIDEr': 1.4070906246554784}
Epoch 14: valid scores: {'BLEU': [0.8408917672561192, 0.697173769630615, 0.5529380543931364, 0.4283263321771929], 'METEOR': 0.30135019234319, 'ROUGE': 0.610824347739308, 'CIDEr': 1.4029420339346297}
Epoch 15: test scores: {'BLEU': [0.8465402268987433, 0.6972535664028359, 0.552456912798488, 0.42814138635150417], 'METEOR': 0.303993079622654, 'ROUGE': 0.6083031533347482, 'CIDEr': 1.4232277540553238}
Epoch 15: valid scores: {'BLEU': [0.8462003193371415, 0.6976896660400096, 0.5517386087777529, 0.4260971025121251], 'METEOR': 0.30345099155722777, 'ROUGE': 0.6092949812007202, 'CIDEr': 1.407506008322609}
Epoch 16: test scores: {'BLEU': [0.8456910780991912, 0.6964113077680676, 0.5512889062739411, 0.4261729272328464], 'METEOR': 0.30301403510052044, 'ROUGE': 0.6075027500710807, 'CIDEr': 1.4254098322216684}
Epoch 16: valid scores: {'BLEU': [0.8459633165319872, 0.697312116592318, 0.5518837873794356, 0.427050827945396], 'METEOR': 0.3033199063556942, 'ROUGE': 0.6093898184649658, 'CIDEr': 1.416808856764353}
Epoch 17: test scores: {'BLEU': [0.8411004815136746, 0.6901909037947318, 0.5457618205353241, 0.42156434462475556], 'METEOR': 0.30523667257339776, 'ROUGE': 0.6066837160033072, 'CIDEr': 1.4334877041755454}
Epoch 17: valid scores: {'BLEU': [0.8410953681936413, 0.692855073956654, 0.5479934232135985, 0.4232043727434191], 'METEOR': 0.30484696139478007, 'ROUGE': 0.6068816711804401, 'CIDEr': 1.4172645559857255}
Epoch 18: test scores: {'BLEU': [0.8403619493437926, 0.6899653182571207, 0.5443875870412581, 0.4202271794567177], 'METEOR': 0.3062425873231951, 'ROUGE': 0.6063952943247476, 'CIDEr': 1.442151582965552}
Epoch 18: valid scores: {'BLEU': [0.8410811487204981, 0.6910268084684723, 0.5444952395260754, 0.4195658558655543], 'METEOR': 0.30481562211534063, 'ROUGE': 0.6062499237604911, 'CIDEr': 1.4287737098441802}
Epoch 19: test scores: {'BLEU': [0.8379903446257846, 0.6886176788284372, 0.5434949946368485, 0.41904074686857645], 'METEOR': 0.30515655394614194, 'ROUGE': 0.605989285204326, 'CIDEr': 1.4371335257986346}
Epoch 19: valid scores: {'BLEU': [0.840395568629888, 0.6941265405518323, 0.5485122612140163, 0.4233698916626582], 'METEOR': 0.3051806795456113, 'ROUGE': 0.6077921017194235, 'CIDEr': 1.4296034286387884}

```
For freezed detector, sorry for the delay. It is located on the server. I may provide you on next Monday.
Hope it helps.
